src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -0,0 +1,285 @@
+
+            return stream;
+        }
+
+        private protected void ReadByteStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<byte> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            var numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+
+            while (!bufferReader.Complete)
+            {
+                var numByteValuesRead = OptimizedByteRLE.ReadValues(ref bufferReader, _byteStreamBuffer);
+
+                for (int idx = 0; idx < numByteValuesRead; idx++)
+                {
+                    if (numSkipped++ < stream.Positions.ValuesToSkip)
+                        continue;
+
+                    outputValues[numValuesRead++] = _byteStreamBuffer[idx];
+
+                    if (numValuesRead >= outputValues.Length)
+                        return;
+                }
+            }
+        }
+
+        private protected void ReadBooleanStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<bool> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            var numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+            var numOfTotalBitsToSkip = stream.Positions.ValuesToSkip * 8 + stream.Positions.RemainingBits;
+            var numOfBytesToSkip = numOfTotalBitsToSkip / 8;
+            while (!bufferReader.Complete)
+            {
+                var numByteValuesRead = OptimizedByteRLE.ReadValues(ref bufferReader, _boolStreamBuffer);
+
+                for (int idx = 0; idx < numByteValuesRead; idx++)
+                {
+                    if (numSkipped++ < numOfBytesToSkip)
+                        continue;
+
+                    var decodedByte = _boolStreamBuffer[idx];
+                    var isFinalByte = bufferReader.Complete && idx >= numByteValuesRead - 1;
+
+                    // Skip remaining bits.
+                    if (numOfBytesToSkip % 8 != 0)
+                        decodedByte = (byte)(decodedByte << numOfTotalBitsToSkip % 8);
+
+                    if (isFinalByte && decodedByte == 0)
+                    {
+                        // Edge case where there is only one value for the row entry and that value is null
+                        outputValues[numValuesRead++] = false;
+                        return;
+                    }
+
+                    for (int bitIdx = 7; bitIdx >= 0; bitIdx--)
+                    {
+                        outputValues[numValuesRead++] = (decodedByte & 1 << bitIdx) != 0;
+
+                        if (numValuesRead >= outputValues.Length)
+                            return;
+
+                        if (isFinalByte && BitOperations.TrailingZeroCount(decodedByte) == bitIdx)
```

Comment:
This is necessary when there are only a few bits set in the last byte, otherwise we would always end up with 8 values. It is the opposite edge case where the last byte is zero (and we have only a single `false` value).

In the C++ implementation, they perform two passes and in the first one they count the number of bits set to discard the ones that may remain from the last byte.

One different approach we can use is to always truncate the `present` stream to the size of the associated data that is being loaded.

Reply:
Yes. The complexity in this case is to make sure the `numValuesRead` will not count more bits than necessary. In tests like the ones we check for a single value, the `finalByte` is  `0b10000000`. We need a way to determine if this is only a single value or if it is a `value` followed by seven `nulls`.

I will double check the possibility to simplify it by always returning the full set of bits (dropping the need for `isFinalByte`). Using the `IntegerColumnBuffer` as example, it would change from this:
```csharp
if (presentValuesRead > 0)
{
    var dataIndex = 0;
    for (int idx = 0; idx < presentValuesRead; idx++)
    {
        if (_presentStreamValues[idx])
            _values[_numValuesRead++] = _dataStreamValues[dataIndex++];
        else
            _values[_numValuesRead++] = null;
    }
}
```
to this (note the change in the `for` loop):
```csharp
if (presentValuesRead > 0)
{
    var dataIndex = 0;
    for (int idx = 0; idx < dataValuesRead; idx++)
    {
        if (_presentStreamValues[idx])
            _values[_numValuesRead++] = _dataStreamValues[dataIndex++];
        else
            _values[_numValuesRead++] = null;
    }
}
```

Reply:
In addition to the example above, lets consider we have two row groups:
- One with 10.000 values, of which 3 are nulls.
- Another with a single value that is not null.

The results from the current implementation would be:

- 10.000 for the `presentValuesRead` and 9.997 for the `dataValuesRead`
- 1 for the `presentValuesRead` and 1 for the `dataValuesRead`

If we REMOVE the `isFinalByte` checks, the result for the second row group would be:

- 8 for the `presentValuesRead` and 1 for the `dataValuesRead`

(cc @ddrinka )